llvm/lib/CAS/HashMappedTrie.cpp

```diff
@@ -194,15 +187,8 @@ struct ThreadSafeHashMappedTrieBase::ImplType {
 
 ThreadSafeHashMappedTrieBase::ImplType &
 ThreadSafeHashMappedTrieBase::getOrCreateImpl() {
-  if (ImplType *Impl = ImplPtr.load())
-    return *Impl;
-
-  // Create a new ImplType and store it if another thread doesn't do so first.
-  // If another thread wins this one is destroyed locally.
-  std::unique_ptr<ImplType> Impl = ImplType::create(0, NumRootBits);
-  ImplType *ExistingImpl = nullptr;
-  (void)ImplPtr.compare_exchange_strong(ExistingImpl, Impl);
-  return *ImplPtr.load();
+  return ImplPtr.loadOrGenerate(
+      [this]() { return ImplType::create(0, NumRootBits); });
```

Comment:
Why not still create it redundantly here (with a normal std::atomic)?

I.e., I had expected:
```
  if (ImplType *Impl = ImplPtr.load())
     return *Impl;

  std::unique_ptr<ImplType> Impl = ImplType::create(0, NumRootBits);
  ImplType *ExistingImpl = nullptr;
  if (ImplPtr.compare_exchange_strong(ExistingImpl, Impl))
    return Impl.release();
  return ExistingImpl;
```

The `loadOrGenerate()` code does look a bit nicer, but LazyAtomicPointer adds the cost of checking for "busy" on every call to `getOrCreateImpl()`. Since the downsides of creating and discarding are pretty low (given that there will be at most one race per data structure) I don't think the extra complexity is warranted.

Reply:
This is actually done in a separate commit in this PR because I feel the code is easier to read this way. If you are worried about the cost of the check, lets go back to the previous implementation with atomic<T*>.

Reply:
Yeah, I think `LazyAtomicPointer` needs to be justified by performance, not just readability. The existing use guards an allocation via bumpptrallocator (which would leak on a race) and arbitrary user code (which we want to guarantee to only call when necessary). There's no potential for a leak here since the ImpPtr isn't bump-ptr-allocated (and never will be); and the extra malloc traffic will be really low since it's once per instance of the (supposed-to-be-long-lived) data structure. So I don't think it's worth it.